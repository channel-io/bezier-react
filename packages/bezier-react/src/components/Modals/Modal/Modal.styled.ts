import { CancelIcon } from '@channel.io/bezier-icons'
import * as DialogPrimitive from '@radix-ui/react-dialog'

import {
  css,
  styled,
} from '~/src/foundation'

import { gap } from '~/src/utils/style'

import {
  Button,
  ButtonColorVariant,
  ButtonSize,
  ButtonStyleVariant,
} from '~/src/components/Button'
import { Text } from '~/src/components/Text'

import {
  type ModalBodyProps,
  type ModalContentProps,
  type ModalFooterProps,
  type ModalHeaderProps,
} from './Modal.types'

import ModalAnimation from './ModalAnimation.styled'

export const DialogPrimitiveOverlay = styled(DialogPrimitive.Overlay)`
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: var(--b-modal-z-index);
  display: grid;
  place-items: center;
  padding: var(--b-modal-collision-padding);
  overflow-y: auto;
  background-color: var(--bgtxt-absolute-black-lighter);

  &[data-state='open'] {
    ${ModalAnimation.overlayShow}
  }
`

export const Content = styled.div<ModalContentProps>`
  ${({ foundation }) => foundation?.rounding.round20}
  ${({ foundation }) => foundation?.elevation.ev4()}

  position: relative;
  box-sizing: border-box;
  width: var(--b-modal-width);
  min-width: 360px;
  max-width: 100vw;
  height: var(--b-modal-height);
  max-height: 100%;
  overflow: visible;
  color: var(--bg-grey-darkest);
  word-break: break-word;

  &[data-state='open'] {
    ${ModalAnimation.contentShow}
  }

  &:focus {
    outline: none;
  }

  /* NOTE(@ed): to show a focus ring when the modal is mounted */
  & [data-component='BezierButton']:focus {
    box-shadow: 0 0 0 3px var(--bgtxt-cobalt-light);
  }

  ${({ interpolation }) => interpolation}
`

export const Section = styled.section`
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 100%;
`

const MODAL_PADDING = 24
const HEADER_BODY_GAP = 12
const FOOTER_TOP_GAP = 16

export const Body = styled.div<ModalBodyProps>`
  flex: 1;
  padding: ${MODAL_PADDING}px;

  ${({ interpolation }) => interpolation}
`

export const Header = styled.header<ModalHeaderProps>`
  display: flex;
  flex-direction: column;
  padding: ${MODAL_PADDING}px ${MODAL_PADDING}px 0 ${MODAL_PADDING}px;
  ${gap(12)}

  /* NOTE(@ed): Support when the ModalBody is used as stand-alone */
  ${({ hidden }) => !hidden && css`
    & + ${Body},
    & + * ${Body} {
      padding-top: ${HEADER_BODY_GAP}px;
    }
  `}

  ${({ interpolation }) => interpolation}
`

export const Footer = styled.footer<ModalFooterProps>`
  display: flex;
  padding: ${FOOTER_TOP_GAP}px ${MODAL_PADDING}px ${MODAL_PADDING}px;

  /* NOTE(@ed): Support when the ModalBody is used as stand-alone */
  /* stylelint-disable declaration-block-semicolon-newline-after, rule-empty-line-before */
  ${Body} + &,
  ${Body} + * & {
    padding-top: 0;
    margin-top: ${FOOTER_TOP_GAP - MODAL_PADDING}px;
  }
  /* stylelint-enable declaration-block-semicolon-newline-after, rule-empty-line-before */

  ${({ interpolation }) => interpolation}
`

export const TitleContainer = styled.div`
  display: flex;
  flex-direction: row;
  width: 100%;
  ${gap(6)}
`

export const HeadingGroup = styled.hgroup`
  display: flex;
  flex: 1;
  flex-direction: column-reverse;
  ${gap(4)}
`

export const Title = styled(Text).attrs({
  forwardedAs: 'h2',
  bold: true,
  color: 'txt-black-darkest',
})`
  width: 100%;
`

const CLOSE_ICON_BUTTON_SIZE = ButtonSize.M
const CLOSE_ICON_BUTTON_MARGIN_X = -6
const CLOSE_ICON_BUTTON_MARGIN_Y = -6

export const CloseIconButton = styled(Button).attrs({
  size: CLOSE_ICON_BUTTON_SIZE,
  leftContent: CancelIcon,
  colorVariant: ButtonColorVariant.MonochromeDark,
  styleVariant: ButtonStyleVariant.Tertiary,
})`
  position: absolute;
  top: ${MODAL_PADDING + CLOSE_ICON_BUTTON_MARGIN_X}px;
  right: ${MODAL_PADDING + CLOSE_ICON_BUTTON_MARGIN_Y}px;
`

export const CloseIconButtonSpacer = styled(Button).attrs({
  forwardedAs: 'div',
  size: CLOSE_ICON_BUTTON_SIZE,
})`
  margin-top: ${CLOSE_ICON_BUTTON_MARGIN_X}px;
  margin-right: ${CLOSE_ICON_BUTTON_MARGIN_Y}px;
  pointer-events: none;
  visibility: hidden;
`

export const FooterLeftContent = styled.div`
  margin-right: auto;
`

export const FooterRightContent = styled.div`
  margin-left: auto;
`
