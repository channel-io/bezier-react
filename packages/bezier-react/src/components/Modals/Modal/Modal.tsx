/* eslint-disable no-restricted-imports */
/* External dependencies */
import React, { forwardRef, useMemo } from 'react'

/* Internal dependencies */
import { BaseModal } from '../BaseModal'
import type { ModalContextValue, ModalProps } from './Modal.types'
import ModalContext from './ModalContext'

function Modal(
  {
    children,
    targetElement,
    onHide,
    ...rests
  }: ModalProps,
  forwardedRef: React.Ref<HTMLDivElement>,
) {
  const contextValue = useMemo<ModalContextValue>(() => ({
    onHide,
  }), [onHide])

  return (
    <BaseModal
      {...rests}
      ref={forwardedRef}
      onHide={onHide}
      targetElement={targetElement}
    >
      <ModalContext.Provider value={contextValue}>
        { children }
      </ModalContext.Provider>
    </BaseModal>
  )
}

export default forwardRef(Modal)
