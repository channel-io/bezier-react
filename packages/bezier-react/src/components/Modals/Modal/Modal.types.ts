import {
  type BezierComponentProps,
  type ChildrenProps,
  type SideContentProps,
} from '~/src/types/ComponentProps'

export enum ModalTitleSize {
  L = 'L',
  M = 'M',
}

interface ModalOptions {
  /**
   * The controlled open state of the modal.
   * Must be used in conjunction with `onShow` and `onHide`.
   */
  show?: boolean

  /**
   * The open state of the modal when it is initially rendered.
   * Use when you **do not need to control** its open state.
   */
  defaultShow?: boolean

  /**
   * Callback function to be called when the modal is opened.
   * @default noop
   */
  onShow?: () => void

  /**
   * Callback function to be called when the modal is closed.
   * @default noop
   */
  onHide?: () => void
}

interface ModalContentOptions {
  /**
   * Specify a container element to portal the content into.
   * @default document.body
   */
  container?: HTMLElement | null

  /**
   * Show close icon button that closes the modal when clicked.
   * @default false
   */
  showCloseIcon?: boolean

  /**
   * Width of the modal.
   * @default 'max-content'
   */
  width?: React.CSSProperties['width']

  /**
   * Height of the modal.
   * @default 'fit-content'
   */
  height?: React.CSSProperties['height']

  /**
   * z-index of the modal content.
   * Rather than using this option, Please check modal is positioned in the proper stacking context.
   * @default ZIndex.Modal
   */
  zIndex?: React.CSSProperties['zIndex']
}

interface ModalHeaderOptions {
  /**
   * An accessible title to be announced when the modal is opened.
   */
  title: React.ReactNode

  /**
   * An accessible subtitle to be announced when the modal is opened.
   */
  subtitle?: React.ReactNode

  /**
   * An accessible description to be announced when the modal is opened.
   */
  description?: React.ReactNode

  /**
   * Size of the title
   * @default ModalTitleSize.L
   */
  titleSize?: ModalTitleSize

  /**
   * Hides content from the screen in an accessible way.
   * @default false
   */
  hidden?: boolean
}

type ModalFooterSideContent = React.ReactNode

interface ModalFooterOptions extends
  SideContentProps<ModalFooterSideContent, ModalFooterSideContent> {}

export interface ModalProps extends
  ChildrenProps,
  ModalOptions {}

export interface ModalContentProps extends
  BezierComponentProps,
  ChildrenProps,
  React.HTMLAttributes<HTMLDivElement>,
  ModalContentOptions {}

export interface ModalHeaderProps extends
  BezierComponentProps,
  Omit<React.HTMLAttributes<HTMLElement>, keyof ModalHeaderOptions>,
  ModalHeaderOptions {}

export interface ModalBodyProps extends
  BezierComponentProps,
  ChildrenProps,
  React.HTMLAttributes<HTMLDivElement> {}

export interface ModalFooterProps extends
  BezierComponentProps,
  React.HTMLAttributes<HTMLElement>,
  ModalFooterOptions {}

export interface ModalTriggerProps extends
  ChildrenProps<React.ReactElement> {}

export interface ModalCloseProps extends
  ChildrenProps<React.ReactElement> {}

export interface ModalContentContextValue extends
  NonNullable<Pick<ModalContentOptions, 'showCloseIcon'>> {}
