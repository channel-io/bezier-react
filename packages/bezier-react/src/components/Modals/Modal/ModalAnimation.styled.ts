/* Internal dependencies */
import { css, keyframes, TransitionDuration } from 'Foundation'

const overlayStyles = {
  open: css`
    opacity: 1;
  `,
  closed: css`
    opacity: 0;
  `,
}

const overlayShow = keyframes`
  from {
    ${overlayStyles.closed}
  }

  to {
    ${overlayStyles.open}
  }
`

const overlayHide = keyframes`
  from {
    ${overlayStyles.open}
  }

  to {
    ${overlayStyles.closed}
  }
`

const contentStyles = {
  open: css`
    opacity: 1;
    transform: translate(-50%, -50%) scale(1);
  `,
  closed: css`
    opacity: 0;
    transform: translate(-50%, -48%) scale(0.95);
  `,
}

const contentShow = keyframes`
  from {
    ${contentStyles.closed}
  }

  to {
    ${contentStyles.open}
  }
`

const contentHide = keyframes`
  from {
    ${contentStyles.open}
  }

  to {
    ${contentStyles.closed}
  }
`

function modalAnimationGenerator(
  frames: ReturnType<typeof keyframes>,
  type: 'show' | 'hide',
) {
  switch (type) {
    case 'show':
      return css`
        animation: ${() => frames} ${TransitionDuration.M}ms cubic-bezier(0.3, 0, 0, 1);
      `
    case 'hide':
      return css`
        animation: ${() => frames} ${TransitionDuration.S}ms cubic-bezier(0.3, 0, 0, 1);
      `
    default:
      return css``
  }
}

export default {
  overlayShow: modalAnimationGenerator(overlayShow, 'show'),
  overlayHide: modalAnimationGenerator(overlayHide, 'hide'),
  contentShow: modalAnimationGenerator(contentShow, 'show'),
  contentHide: modalAnimationGenerator(contentHide, 'hide'),
}
