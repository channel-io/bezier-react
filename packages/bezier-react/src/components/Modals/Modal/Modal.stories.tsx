/* eslint-disable max-len */
/* External dependencies */
import React from 'react'
import base from 'paths.macro'
import { Story, Meta } from '@storybook/react'

/* Internal dependencies */
import { getTitle } from 'Utils/storyUtils'
import {
  ModalProps,
  ModalContentProps,
  ModalActionProps,
  ModalTitleSize,
} from './Modal.types'
import Modal from './Modal'
import ModalContent from './ModalContent'
import ModalAction from './ModalAction'

export default {
  title: getTitle(base),
  component: Modal,
  argTypes: {
    titleSize: {
      control: {
        type: 'radio',
        options: [
          ...Object.values(ModalTitleSize),
        ],
      },
    },
  },
} as Meta

interface ModalStorybookProps extends
  Omit<ModalProps, 'title'>,
  ModalContentProps,
  ModalActionProps { }

const DEFAULT_OPTIONAL_MODAL_PROPS: Partial<ModalStorybookProps> = {
  show: false,
  autoFocus: true,
  showCloseIcon: false,
  padded: true,
  backdropClassName: '',
  targetElement: undefined,
  zIndex: 1e7,
}

const ModalStorybook = ({
  title,
  titleSize,
  subTitle,
  description,
  showCloseIcon,
  leftContent,
  rightContent,
  show,
  children,
  ...rests
}: ModalStorybookProps) => {
  const [isShow, setShow] = React.useState(show)

  React.useEffect(function watchShowToChange() {
    setShow(show)
  }, [show])

  const onHide = React.useCallback(() => {
    setShow(false)
  }, [])

  const handleShowAsToggle = React.useCallback(() => {
    setShow(!isShow)
  }, [isShow])

  return (
    <>
      <button
        type="button"
        onClick={handleShowAsToggle}
      >
        { isShow ? 'Hide' : 'Show' }
      </button>
      <Modal
        {...rests}
        show={isShow}
        onHide={onHide}
      >
        <ModalContent
          title={title}
          titleSize={titleSize}
          subTitle={subTitle}
          description={description}
          showCloseIcon={showCloseIcon}
        >
          { children }
        </ModalContent>
        <ModalAction
          leftContent={leftContent}
          rightContent={rightContent}
        />
      </Modal>
    </>
  )
}

const PrimaryTemplate: Story<ModalStorybookProps> = (props) => (
  <ModalStorybook {...props} />
)

export const Primary = PrimaryTemplate.bind({})
Primary.args = {
  ...DEFAULT_OPTIONAL_MODAL_PROPS,
  title: 'Title',
  description: 'Description',
  leftContent: 'Left content',
  rightContent: 'Right content',
}

export const WithChildren = PrimaryTemplate.bind({})
WithChildren.args = {
  ...DEFAULT_OPTIONAL_MODAL_PROPS,
  title: 'Title',
  subTitle: 'Sub Title',
  description: 'Description',
  leftContent: 'Left content',
  rightContent: 'Right content',
  children: (
    <div style={{ width: '100%', height: '50px', backgroundColor: 'rgba(94, 86, 240, 0.1)' }} />
  ),
}

export const WithSubTitle = PrimaryTemplate.bind({})
WithSubTitle.args = {
  ...DEFAULT_OPTIONAL_MODAL_PROPS,
  title: 'Title',
  subTitle: 'Sub Title',
  description: 'Description',
  leftContent: 'Left content',
  rightContent: 'Right content',
}

export const LongTitle = PrimaryTemplate.bind({})
LongTitle.args = {
  ...DEFAULT_OPTIONAL_MODAL_PROPS,
  title: Array.from(Array(500)).map(() => 'Title ').join(''),
  description: 'Description',
  leftContent: 'Left content',
  rightContent: 'Right content',
}

export const LongDescription = PrimaryTemplate.bind({})
LongDescription.args = {
  ...DEFAULT_OPTIONAL_MODAL_PROPS,
  title: 'Title',
  description: Array.from(Array(500)).map(() => 'Description ').join(''),
  leftContent: 'Left content',
  rightContent: 'Right content',
}

export const LongTitleAndDescription = PrimaryTemplate.bind({})
LongTitleAndDescription.args = {
  ...DEFAULT_OPTIONAL_MODAL_PROPS,
  title: Array.from(Array(500)).map(() => 'Title ').join(''),
  description: Array.from(Array(500)).map(() => 'Description ').join(''),
  leftContent: 'Left content',
  rightContent: 'Right content',
}

export const WithoutTitle = PrimaryTemplate.bind({})
WithoutTitle.args = {
  ...DEFAULT_OPTIONAL_MODAL_PROPS,
  description: 'Description',
  leftContent: 'Left content',
  rightContent: 'Right content',
}

export const WithoutDescription = PrimaryTemplate.bind({})
WithoutDescription.args = {
  ...DEFAULT_OPTIONAL_MODAL_PROPS,
  title: 'Title',
  leftContent: 'Left content',
  rightContent: 'Right content',
}

export const WithoutAction = PrimaryTemplate.bind({})
WithoutAction.args = {
  ...DEFAULT_OPTIONAL_MODAL_PROPS,
  title: 'Title',
  description: 'Description',
}

