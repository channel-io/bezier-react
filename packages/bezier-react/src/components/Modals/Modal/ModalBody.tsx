/* External dependencies */
import React, { forwardRef } from 'react'

/* Internal dependencies */
import { type ModalBodyProps } from './Modal.types'
import * as Styled from './Modal.styled'

/**
 * `ModalBody` is a simple wrapper of the main modal content.
 */
export const ModalBody = forwardRef(function ModalBody({
  children,
  ...rest
}: ModalBodyProps, forwardedRef: React.Ref<HTMLDivElement>) {
  return (
    <Styled.Body
      ref={forwardedRef}
      {...rest}
    >
      { children }
    </Styled.Body>
  )
})
