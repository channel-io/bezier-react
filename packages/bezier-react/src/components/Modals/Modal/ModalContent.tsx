/* External dependencies */
import React, { forwardRef, useMemo } from 'react'
import * as DialogPrimitive from '@radix-ui/react-dialog'
import { isNumber } from 'lodash-es'

/* Internal dependencies */
import { document } from 'Utils/domUtils'
import { ModalClose } from './ModalHelpers'
import ModalContentContext from './ModalContentContext'
import { ModalContentProps, ModalContentContextValue } from './Modal.types'
import * as Styled from './Modal.styled'

/**
 * `ModalContent` is a container of the modal content.
 * It creates a portal to render the modal content outside of the DOM tree
 * and renders overlay behind the modal content too.
 */
export const ModalContent = forwardRef(function ModalContent({
  children,
  style,
  container = document.body,
  showCloseIcon = false,
  width = 'max-content',
  height = 'fit-content',
  zIndex = 0,
  ...rest
}: ModalContentProps, forwardedRef: React.Ref<HTMLDivElement>) {
  const overlayStyle = useMemo((): React.CSSProperties & {
    '--bezier-modal-z-index': ModalContentProps['zIndex']
  } => ({
    '--bezier-modal-z-index': zIndex,
  }), [zIndex])

  const contentStyle = useMemo((): React.CSSProperties & {
    '--bezier-modal-width': ModalContentProps['width']
    '--bezier-modal-height': ModalContentProps['height']
  } => ({
    ...style,
    '--bezier-modal-width': isNumber(width) ? `${width}px` : width,
    '--bezier-modal-height': isNumber(height) ? `${height}px` : height,
  }), [
    style,
    width,
    height,
  ])

  const contextValue = useMemo((): ModalContentContextValue => ({
    showCloseIcon,
  }), [showCloseIcon])

  return (
    <DialogPrimitive.Portal container={container}>
      <Styled.DialogPrimitiveOverlay style={overlayStyle}>
        <DialogPrimitive.Content asChild>
          <Styled.Content
            aria-modal
            ref={forwardedRef}
            style={contentStyle}
            {...rest}
          >
            <Styled.Section>
              <ModalContentContext.Provider value={contextValue}>
                { children }
              </ModalContentContext.Provider>

              { /* NOTE: To prevent focusing first on the close button when opening the modal, place the close button behind. */ }
              { showCloseIcon && (
                <ModalClose>
                  <Styled.CloseIconButton />
                </ModalClose>
              ) }
            </Styled.Section>
          </Styled.Content>
        </DialogPrimitive.Content>
      </Styled.DialogPrimitiveOverlay>
    </DialogPrimitive.Portal>
  )
})
