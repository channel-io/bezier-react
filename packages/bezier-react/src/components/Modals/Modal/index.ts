import Modal from './Modal'
import ModalContent from './ModalContent'
import ModalAction from './ModalAction'

import type {
  ModalProps,
  ModalActionProps,
  ModalContentProps,
  ModalContextValue,
} from './Modal.types'

export type {
  ModalProps,
  ModalActionProps,
  ModalContentProps,
  ModalContextValue,
}
export {
  Modal,
  ModalContent,
  ModalAction,
}
