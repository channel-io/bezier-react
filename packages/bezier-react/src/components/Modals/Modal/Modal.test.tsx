/* External dependencies */
import React from 'react'
import type { RenderResult } from '@testing-library/react'
import { fireEvent } from '@testing-library/react'

/* Internal dependencies */
import { render } from 'Utils/testUtils'
import type {
  ModalProps,
  ModalContentProps,
  ModalActionProps,
} from './Modal.types'
import Modal from './Modal'
import ModalContent from './ModalContent'
import ModalAction from './ModalAction'

const DEFAULT_PROPS: ModalProps = {
  show: true,
  className: 'Modal',
  onHide: jest.fn(),
}

interface ModalTestRenderProps extends
  Omit<ModalProps, 'title'>,
  ModalContentProps,
  ModalActionProps { }

const ModalTestRender = ({
  title,
  subTitle,
  description,
  showCloseIcon,
  leftContent,
  rightContent,
  children,
  ...rests
}: ModalTestRenderProps) => (
  <Modal {...rests}>
    <ModalContent
      title={title}
      subTitle={subTitle}
      description={description}
      showCloseIcon={showCloseIcon}
    >
      { children }
    </ModalContent>
    <ModalAction
      leftContent={leftContent}
      rightContent={rightContent}
    />
  </Modal>
)

const main = document.createElement('div')
main.setAttribute('id', 'main')

describe('Modal', () => {
  let renderResult: RenderResult

  describe('When "show" is false', () => {
    beforeEach(() => {
      renderResult = render(
        <ModalTestRender
          {...DEFAULT_PROPS}
          show={false}
        />,
        {
          container: document.body.appendChild(main),
        },
      )
    })

    afterEach(jest.clearAllMocks)

    it('Component is "null"', async () => {
      const target = renderResult.container.firstChild
      expect(target).toBe(null)
    })
  })

  describe('When "show" is true', () => {
    beforeEach(() => {
      renderResult = render(
        <ModalTestRender
          {...DEFAULT_PROPS}
          className="Modal__Contests__Test"
        />,
        {
          container: document.body.appendChild(main),
        },
      )
    })

    afterEach(jest.clearAllMocks)

    it('should equal the snapshot', async () => {
      const target = await renderResult.findByTestId('BaseModal')
      expect(target).toMatchSnapshot()
    })

    describe('Modal__Contents', () => {
      beforeEach(() => {
        renderResult = render(
          <ModalTestRender
            {...DEFAULT_PROPS}
            title="Title"
            subTitle="SubTitle"
            description="Description"
            leftContent="Left content"
            rightContent="Right content"
          />,
          {
            container: document.body.appendChild(main),
          },
        )
      })

      afterEach(jest.clearAllMocks)

      it('Styles', async () => {
        const target = await renderResult.findByTestId('Modal__Contents')
        expect(target).toHaveStyle('box-sizing: border-box;')
        expect(target).toHaveStyle('width: 100%;')
        expect(target).toHaveStyle('padding: 24px 24px 16px;')
      })

      describe('Modal__Contents__Title', () => {
        describe('When "title" and "subTitle" are nil', () => {
          beforeEach(() => {
            renderResult = render(
              <ModalTestRender
                {...DEFAULT_PROPS}
                description="Description"
                leftContent="Left content"
                rightContent="Right content"
                showCloseIcon
              />,
              {
                container: document.body.appendChild(main),
              },
            )
          })

          it('Modal__Contents__Title is null', () => {
            const target = renderResult.queryByTestId('Modal__Contents__Title')
            expect(target).toBe(null)
          })
        })

        beforeEach(() => {
          renderResult = render(
            <ModalTestRender
              {...DEFAULT_PROPS}
              title="Title"
              subTitle="SubTitle"
              description="Description"
              leftContent="Left content"
              rightContent="Right content"
            />,
            {
              container: document.body.appendChild(main),
            },
          )
        })

        afterEach(jest.clearAllMocks)

        it('Styles', async () => {
          const target = await renderResult.findByTestId('Modal__Contents__Title')
          expect(target).toHaveStyle('display: flex;')
          expect(target).toHaveStyle('flex: 1;')
        })

        describe('Modal__Contents__TitleAndSubTitle', () => {
          it('Styles', async () => {
            const target = await renderResult.findByTestId('Modal__Contents__TitleAndSubTitle')
            expect(target).toHaveStyle('display: block;')
            expect(target).toHaveStyle('width: 100%;')
          })

          describe('Modal__Contents__Title__Text', () => {
            it('Styles', async () => {
              const target = await renderResult.findByTestId('Modal__Contents__Title__Text')
              expect(target).toHaveStyle('box-sizing: border-box;')
              expect(target).toHaveStyle('display: block;')
              expect(target).toHaveStyle('width: 100%;')
              expect(target).toHaveStyle('overflow: hidden;')
              expect(target).toHaveStyle('text-overflow: ellipsis;')
              expect(target).toHaveStyle('white-space: nowrap;')
            })
          })

          describe('Modal__Contents__SubTitle__Text', () => {
            it('Styles', async () => {
              const target = await renderResult.findByTestId('Modal__Contents__SubTitle__Text')
              expect(target).toHaveStyle('box-sizing: border-box;')
              expect(target).toHaveStyle('display: block;')
              expect(target).toHaveStyle('width: 100%;')
              expect(target).toHaveStyle('overflow: hidden;')
              expect(target).toHaveStyle('text-overflow: ellipsis;')
              expect(target).toHaveStyle('white-space: nowrap;')
            })
          })
        })

        describe('Modal__Contents__Title__CloseButton', () => {
          describe('When "showCloseIcon" is true', () => {
            beforeEach(() => {
              renderResult = render(
                <ModalTestRender
                  {...DEFAULT_PROPS}
                  title="Title"
                  subTitle="SubTitle"
                  description="Description"
                  leftContent="Left content"
                  rightContent="Right content"
                  showCloseIcon
                />,
                {
                  container: document.body.appendChild(main),
                },
              )
            })

            it('CloseButton is mounted', async () => {
              const target = await renderResult.findByTestId('Modal__Contents__Title__CloseButton')
              expect(target).toBeInTheDocument()
            })

            it('CloseButton is clicked, the modal is closed.', async () => {
              const target = await renderResult.findByTestId('Modal__Contents__Title__CloseButton')
              fireEvent.click(target)
              expect(DEFAULT_PROPS.onHide).toBeCalledTimes(1)
            })

            it('Styles', async () => {
              const target = await renderResult.findByTestId('Modal__Contents__Title__CloseButton')
              expect(target).toHaveStyle('display: inline-block;')
              expect(target).toHaveStyle('background-color: transparent;')
              expect(target).toHaveStyle('margin-top: -6px;')
              expect(target).toHaveStyle('margin-right: -6px;')
            })
          })
        })
      })
    })

    describe('Modal__Action', () => {
      describe('When "leftContent" and "rightContent" are nil', () => {
        beforeEach(() => {
          renderResult = render(
            <ModalTestRender
              {...DEFAULT_PROPS}
              title="Title"
              subTitle="SubTitle"
              description="Description"
              showCloseIcon
            />,
            {
              container: document.body.appendChild(main),
            },
          )
        })

        it('Modal__Action is null', () => {
          const target = renderResult.queryByTestId('Modal__Action')
          expect(target).toBe(null)
        })
      })

      beforeEach(() => {
        renderResult = render(
          <ModalTestRender
            {...DEFAULT_PROPS}
            title="Title"
            subTitle="SubTitle"
            description="Description"
            leftContent="Left content"
            rightContent="Right content"
          />,
          {
            container: document.body.appendChild(main),

          },
        )
      })

      afterEach(jest.clearAllMocks)

      it('Styles', async () => {
        const target = await renderResult.findByTestId('Modal__Action')
        expect(target).toHaveStyle('box-sizing: border-box;')
        expect(target).toHaveStyle('display: flex')
        expect(target).toHaveStyle('flex: 1')
        expect(target).toHaveStyle('justify-content: space-between')
        expect(target).toHaveStyle('padding: 0 24px 24px')
      })
    })
  })
})
