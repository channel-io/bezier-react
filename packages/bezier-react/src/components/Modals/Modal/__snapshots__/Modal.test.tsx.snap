// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal When "show" is true should equal the snapshot 1`] = `
.c0 {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 40px 0;
  -webkit-animation-name: jgQpwH;
  animation-name: jgQpwH;
  -webkit-animation-duration: 300ms;
  animation-duration: 300ms;
}

.c1 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: #00000066;
}

.c2 {
  position: relative;
  z-index: 1;
  min-width: 360px;
  max-height: 100%;
  overflow-y: auto;
  color: #242428;
  background-color: #FFFFFF;
  border-radius: 20px;
  -webkit-animation-name: gUeayS;
  animation-name: gUeayS;
  -webkit-animation-duration: 300ms;
  animation-duration: 300ms;
  background-color: #FFFFFF;
  box-shadow: inset 0 0 2px 0 #FFFFFF1F, 0 0 2px 1px #0000000D, 0 4px 20px #00000038;
}

.c3 {
  box-sizing: border-box;
  width: 100%;
  padding: 24px 24px 16px;
}

<div
  class="c0"
  data-testid="BaseModal"
  style="z-index: 10000000;"
>
  <div
    class="c1"
    data-testid="BaseModal__Background"
  />
  <div
    class="c2 Modal__Contests__Test"
    data-testid="BaseModal__Children"
    tabindex="-1"
  >
    <div
      class="c3"
      data-testid="Modal__Contents"
    />
  </div>
</div>
`;
